### ISSUE TITLE
segfault in pth_mutex_release in shadow thread

### ISSUE THREAD

**Issue by dhuseby** [shadow/shadow#402]
```
Thread 1 "shadow" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7876b92f3ea0 (LWP 27191)]
0x00007876b172adcf in ?? ()
(gdb) bt
#0  0x00007876b172adcf in ?? ()
#1  0x000061e340240e00 in ?? ()
#2  0x000000000000000f in ?? ()
#3  0x00007876b0acac48 in ?? ()
#4  0x00007876b9133c04 in get_current_dtv (tp=<optimized out>) at /home/user/Projects/shadow/src/external/elf-loader/vdl-tls.c:306
#5  vdl_tls_get_addr_fast (module=132451165579872, offset=132451307981912) at /home/user/Projects/shadow/src/external/elf-loader/vdl-tls.c:615
#6  0x00007876b0acab98 in ?? ()
#7  0x00007876b0acab40 in ?? ()
#8  0x00007876b17376e2 in ?? ()
#9  0x000061e34023bb30 in ?? ()
#10 0x000061e33ece2e7b in pth_mutex_release (mutex=0x61e3402420d0) at /home/user/Projects/shadow/src/external/rpth/pth_sync.c:122
#11 0x0000000000000019 in ?? ()
#12 0x0000000000000000 in ?? ()
(gdb) info thread
  Id   Target Id         Frame 
* 1    Thread 0x7876b92f3ea0 (LWP 27191) "shadow" 0x00007876b172adcf in ?? ()
  2    Thread 0x7876b23ce700 (LWP 27193) "logger-helper" syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
(gdb) info registers
rax            0x7876b9298c58   132451307981912
rbx            0x7876b9298c58   132451307981912
rcx            0x3      3
rdx            0x7876b933c948   132451308652872
rsi            0x0      0
rdi            0x0      0
rbp            0x7876b0acaa60   0x7876b0acaa60
rsp            0x7876b0acaa10   0x7876b0acaa10
r8             0x0      0
r9             0x30     48
r10            0x30     48
r11            0x202    514
r12            0x7876b9298c58   132451307981912
r13            0x19     25
r14            0x7876b0acab98   132451165580184
r15            0xd      13
rip            0x7876b172adcf   0x7876b172adcf
eflags         0x10246  [ PF ZF IF RF ]
cs             0xe033   57395
ss             0xe02b   57387
ds             0x0      0
es             0x0      0
fs             0x0      0
gs             0x0      0
(gdb) 
```

TL;DR

I'm running the shadow master branch revision 3ba48737 and [https://github.com/dhuseby/shadow-rust-test](https://github.com/dhuseby/shadow-rust-test) revision 4d4f6d9.  I'm running shadow using the shadow-rust-test.xml config in the shadow-rust-test repo.  The shadow thread segfaults somewhere in get_current_dtv in the elf loader.  I haven't had time to really dig into it yet.  Just wanted to put it here first.

**Comment by jtracey** [comment]
I still need to document this, but with some recent commits debug symbols for plugins aren't loaded by default. Since this experiment is small, you can get away with running `p vdl_linkmap_abi_update()` in gdb after a segfault and it will correct everything.

I'm getting a segfault as well, in
`core::ptr::swap_nonoverlapping_bytes () at /checkout/src/libcore/ptr.rs:187`
with the assembly instruction
`movdqa 0x80(%rax),%xmm0`
Where %rax is some valid address that isn't aligned (movqda assumes the argument is 16 byte aligned and segfaults when it's not). The place where I've seen this most commonly in the past is a misaligned stack, but I *think* the stack is still aligned here, so I don't know what's causing it. It doesn't look like it was any recent Shadow change that caused this, do you know which change in particular this started happening with?

**Comment by dhuseby** [comment]
I think this was probably always there and that now that the earlier rust related fixes are in, we're now seeing this one.
